adding a test for ddp save()/load()

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44906 adding a test for ddp save()/load()**
* #44905 renaming TestDdpCommHook class so it doesn't get picked up as a test by pytest

Differential Revision: [D23825386](https://our.internmc.facebook.com/intern/diff/D23825386)

bdhirsh committed:
adding a test for ddp save()/load()

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"




[ghstack-poisoned]

mrshenli reviewed (commented):
LGTM. I added some nit comments. The lint failure is real though.

bdhirsh committed:
Update on "adding a test for ddp save()/load()"




[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"




[ghstack-poisoned]

mrshenli reviewed (changes requested):
Test failure is real:

```
Sep 18 17:28:26   test_save_load_checkpoint (__main__.DistributedDataParallelTest) ... ERROR:root:Caught exception: 
Sep 18 17:28:26 Traceback (most recent call last):
Sep 18 17:28:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 246, in wrapper
Sep 18 17:28:26     fn()
Sep 18 17:28:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 67, in wrapper
Sep 18 17:28:26     return func(*args, **kwargs)
Sep 18 17:28:26   File "distributed/test_c10d.py", line 2948, in test_save_load_checkpoint
Sep 18 17:28:26     optimizer_withload = optim.SGD(ddp_withload.parameters(), lr=0.001)
Sep 18 17:28:26 NameError: name 'optim' is not defined
Sep 18 17:28:26 exiting process with exit code: 10
Sep 18 17:28:26 ERROR:root:Caught exception: 
Sep 18 17:28:26 Traceback (most recent call last):
Sep 18 17:28:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 246, in wrapper
Sep 18 17:28:26     fn()
Sep 18 17:28:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 67, in wrapper
Sep 18 17:28:26     return func(*args, **kwargs)
Sep 18 17:28:26   File "distributed/test_c10d.py", line 2948, in test_save_load_checkpoint
Sep 18 17:28:26     optimizer_withload = optim.SGD(ddp_withload.parameters(), lr=0.001)
Sep 18 17:28:26 NameError: name 'optim' is not defined
Sep 18 17:28:26 exiting process with exit code: 10
```

bdhirsh committed:
Update on "adding a test for ddp save()/load()"




[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"




[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks for adding this!

bdhirsh committed:
Update on "adding a test for ddp save()/load()"




[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -2900,6 +2900,80 @@ def forward(self, x):
+        checkpoint_path = self.file_name + "_model_checkpoint"
+        if self.rank == 0:
+            torch.save(ddp_withload.state_dict(), checkpoint_path)
+        dist.barrier()
```
Since `init_process_group` was not called, the default pg is not available. Can you try calling `process_group.barrier()` instead of `dist.barrier()`?

bdhirsh committed:
Update on "adding a test for ddp save()/load()"


Differential Revision: [D23825386](https://our.internmc.facebook.com/intern/diff/D23825386)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"


Differential Revision: [D23825386](https://our.internmc.facebook.com/intern/diff/D23825386)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"


Differential Revision: [D23825386](https://our.internmc.facebook.com/intern/diff/D23825386)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"


Differential Revision: [D23825386](https://our.internmc.facebook.com/intern/diff/D23825386)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a test for ddp save()/load()"


Differential Revision: [D23825386](https://our.internmc.facebook.com/intern/diff/D23825386)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Thanks for adding this test!
